aten/src/ATen/native/TensorIndexing.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <ATen/core/TensorBody.h>
```

Comment:
TODO: have a top level comment on this header explaining the general strategy here.